File: Low Level Design (LLD)/DesignPatterns/03. Singleton/CPP/ByReference (Thread Safe)/Logger.cpp
#include "./Logger.h"

int Logger::counter = 0;

Logger::Logger() {
    counter++;
    std::cout << "Instance created, no of instance: " << counter << std::endl;
}

Logger& Logger::getLogger() {
    static Logger loggerInstance;
    return loggerInstance;
}

void Logger::Log(std::string message) {
    std::cout << message << std::endl;
}

File: Low Level Design (LLD)/DesignPatterns/03. Singleton/Java/Logger.java
public class Logger {

    private static Logger loggerInstance = null;
    private static int counter = 0;

    private Logger() {
        counter++;
        System.out.println("Instance created, number of instances: " + counter);
    }

    public static synchronized Logger getLogger() {
        if (loggerInstance == null) {
            loggerInstance = new Logger();
        }
        return loggerInstance;
    }

    public void Log(String message) {
        System.out.println(message);
    }
}
